src/lib.rs

```diff
@@ -0,0 +1,371 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `task-def` - Task definition of fuzzer to generate.
+    /// * `build_variant` -
+    /// * `config_location` - Location where generated configuration will be stored in S3.
+    ///
+    /// # Returns
+    ///
+    /// Parameters to configure how fuzzer task should be generated.
+    fn task_def_to_fuzzer_params(
+        &self,
+        task_def: &EvgTask,
+        build_variant: &BuildVariant,
+        config_location: &str,
+    ) -> Result<FuzzerGenTaskParams>;
+}
+
+struct GenerateTasksServiceImpl {
+    evg_config_service: Arc<dyn EvgConfigService>,
+    evg_config_utils: Arc<dyn EvgConfigUtils>,
+    gen_fuzzer_service: Arc<dyn GenFuzzerService>,
+}
+
+impl GenerateTasksServiceImpl {
+    /// Create an instance of GenerateTasksServiceImpl.
+    ///
+    /// # Arguments
+    ///
+    /// * `evg_config_service` - Service to work with evergreen project configuration.
+    /// * `evg_config_utils` - Utilities to work with evergreen project configuration.
+    /// * `gen_fuzzer_service` - Service to generate fuzzer tasks.
+    pub fn new(
+        evg_config_service: Arc<dyn EvgConfigService>,
+        evg_config_utils: Arc<dyn EvgConfigUtils>,
+        gen_fuzzer_service: Arc<dyn GenFuzzerService>,
+    ) -> Self {
+        Self {
+            evg_config_service,
+            evg_config_utils,
+            gen_fuzzer_service,
+        }
+    }
+}
+
+/// An implementation of GeneratorTasksService.
+impl GenerateTasksService for GenerateTasksServiceImpl {
+    /// Build task definition for all tasks
+    ///
+    /// # Arguments
+    ///
+    /// * `config_location` - Location in S3 where generated configuration will be stored.
+    ///
+    /// Returns
+    ///
+    /// Map of task names to generated task definitions.
+    fn build_generated_tasks(
+        &self,
+        config_location: &str,
+    ) -> Result<Arc<Mutex<GenTaskCollection>>> {
+        let build_variant_list = self.evg_config_service.sort_build_variants_by_required();
+        let build_variant_map = self.evg_config_service.get_build_variant_map();
+        let task_map = self.evg_config_service.get_task_def_map();
+
+        let generated_tasks = Arc::new(Mutex::new(HashMap::new()));
+        let mut seen_tasks = HashSet::new();
+        for build_variant in &build_variant_list {
+            let build_variant = build_variant_map.get(build_variant).unwrap();
+            for task in &build_variant.tasks {
+                if !seen_tasks.contains(&task.name) {
+                    seen_tasks.insert(task.name.to_string());
+                    if let Some(task_def) = task_map.get(&task.name) {
+                        if self.evg_config_utils.is_task_generated(task_def) {
+                            if self.evg_config_utils.is_task_fuzzer(task_def) {
+                                event!(Level::INFO, "Generating fuzzer: {}", &task.name,);
```

Comment:
This is pretty deep nesting and makes it a bit hard to understand whats happening here. Is there anything we can do to reduce how nested the interesting code portions are?

Reply:
I broke up this a little and did some refactoring to reduce the nesting some. Hopefully that makes it a little easier to follow.

One think to note, the next ticket (adding support for generating resmoke tasks) will have a pretty big impact on this section of the code. So there are likely more changes coming in the future.